src/GazeboUnderwater.cpp

```diff
@@ -28,74 +27,133 @@ namespace gazebo_underwater
     {
+        // Test if model_name is defined in the world file
         if(sdf->HasElement("model_name"))
         {
+            // Test if model_name name a model loaded in gazebo
             model = world->GetModel( sdf->Get<std::string>("model_name") );  
-            gzmsg << "GazeboUnderwater: found model:" << model->GetName() << std::endl;
+            if(model)
+            {
+                gzmsg << "GazeboUnderwater: model: " << model->GetName() << std::endl;
+            }else{
+                std::string msg = "GazeboUnderwater: model " + sdf->Get<std::string>("model_name")
+                        + " not found in gazebo world " + world->GetName();
+                gzthrow(msg);
+            }
         }else{
-            gzmsg << "GazeboUnderwater: model not found. Quit simulation... " << std::endl;
+            gzthrow("GazeboUnderwater: model_name not defined in world file !");
+        }
+
+        if(sdf->HasElement("link_name"))
+        {
+            link = model->GetLink( sdf->Get<std::string>("link_name") );
+            if( link ){
+                gzmsg << "GazeboUnderwater: link: " << link->GetName() << std::endl;
+                physics::InertialPtr modelInertia = link->GetInertial();
+                gzmsg << "GazeboUnderwater: link mass: " << modelInertia->GetMass() << std::endl;
+            }else{
+                std::string msg = "GazeboUnderwater: link " + sdf->Get<std::string>("link_name")
+                        + " not found in model " + model->GetName();
+                gzthrow(msg);
+            }
+        }else{
+            gzthrow("GazeboUnderwater: link_name not defined in world file !");
         }
 
-        size = getParameter<math::Vector3>("size","meters",
-                math::Vector3(1, 1, 1));
-        centerOfBuoyancy = getParameter<math::Vector3>("center_of_buoyancy","meters",
-                math::Vector3(0, 0, 0.1));
-        fluidVelocity = getParameter<math::Vector3>("fluid_velocity","m/s",
-                math::Vector3(0,0,0));
-        viscousDamping = getParameter<math::Vector3>("viscous_damping","dimensionless",
-                math::Vector3(20,30,50));
         waterLevel = getParameter<double>("water_level","meters", 2.0);
-        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1000);
+        fluidVelocity = getParameter<math::Vector3>("fluid_velocity","m/s",math::Vector3(0,0,0));
+        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1027);
+        dragCoefficient = getParameter<math::Vector3>("drag_coefficient","dimensionless",
+                math::Vector3(1,1,1));
+        // buoyancy must be the buoyancy when the model is completely submersed
+        buoyancy = getParameter<double>("buoyancy","N",0);
+        centerOfBuoyancy = getParameter<math::Vector3>("center_of_buoyancy","meters",
+                math::Vector3(0, 0, 0.15));
+        sideAreas = getParameter<math::Vector3>("side_areas","meter2",
```

Comment:
Am 01.04.2015 um 17:33 schrieb Thomio Watanabe <notifications@github.com>:
> In src/GazeboUnderwater.cpp:
> 
> >          waterLevel = getParameter<double>("water_level","meters", 2.0);
> > -        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1000);
> > +        fluidVelocity = getParameter<math::Vector3>("fluid_velocity","m/s",math::Vector3(0,0,0));
> > +        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1027);
> > +        dragCoefficient = getParameter<math::Vector3>("drag_coefficient","dimensionless",
> > +                math::Vector3(1,1,1));
> > +        // buoyancy must be the buoyancy when the model is completely submersed
> > +        buoyancy = getParameter<double>("buoyancy","N",0);
> > +        centerOfBuoyancy = getParameter<math::Vector3>("center_of_buoyancy","meters",
> > +                math::Vector3(0, 0, 0.15));
> > +        sideAreas = getParameter<math::Vector3>("side_areas","meter2",
> 
> I was speaking with Jan and I think he wants this way.
> The same for the volume variable. (because the real robot may have empty spaces inside it)

Using the bounding box for the calculation only makes sense, if the volume of this box actually represents the volume of water displaced. In most underwater systems that is not entirely the case, especially with a covered open frame like FlatFish (or HROV for that matter).

So I thought its better to force the user to provide displaced Volume and weight as a parameter.

blue skies
	Jan

--
-- Jan Christian Albiez -- jan@ankerwin.de -- http://www.ankerwin.de/ --
-- segmentation violation in module reality.o ...
                         ... please shutdown your universe and reboot --